drivers/serial/uart_nrfx_uarte.c

```diff
@@ -915,33 +931,31 @@ static void uarte_nrfx_poll_out(struct device *dev, unsigned char c)
 				uarte_nrfx_isr_async(dev);
 			}
 		}
-		/* Set tx_size but don't set tx_buf to differentiate this
-		 * transmission from the one started with uarte_nrfx_tx,
-		 * this way uarte_nrfx_tx will return -EBUSY and poll out will
-		 * work when interrupted.
+		/* Use tx_size as lock, this way uarte_nrfx_tx will
+		 * return -EBUSY during poll_out.
 		 */
-		data->async->tx_size = 1;
-		nrf_uarte_int_disable(uarte,
-				      NRF_UARTE_INT_ENDTX_MASK |
-				      NRF_UARTE_INT_TXSTOPPED_MASK);
-	}
+		lock = &data->async->tx_size;
+	} else
 #endif
-	/* The UART API dictates that poll_out should wait for the transmitter
-	 * to be empty before sending a character. However, the only way of
-	 * telling if the transmitter became empty in the UARTE peripheral is
-	 * to check if the ENDTX event for the previous transmission was set.
-	 * Since this event is not cleared automatically when a new transmission
-	 * is started, it must be cleared in software, and this leads to a rare
-	 * yet possible race condition if the thread is preempted right after
-	 * clearing the event but before sending a new character. The preempting
-	 * thread, if it also called poll_out, would then wait for the ENDTX
-	 * event that had no chance to become set.
-
-	 * Because of this race condition, the while loop has to be placed
-	 * after the write to TXD, and we can't wait for an empty transmitter
-	 * before writing. This is a trade-off between losing a byte once in a
-	 * blue moon against hanging up the whole thread permanently
-	 */
+		lock = &data->poll_out_lock;
+
+	if (!k_is_in_isr()) {
+		u8_t safety_cnt = 100;
+
+		while (atomic_cas((atomic_t *) lock,
+				(atomic_val_t) 0,
+				(atomic_val_t) 1) == false) {
+			/* k_sleep allows other threads to execute and finish
+			 * their transactions.
+			 */
+			k_sleep(1);
+			if (--safety_cnt == 0) {
+				break;
```

Comment:
I guess it would be safer to return than to continue without obtaining the lock.

Reply:
We end up here if current thread is using poll out while lower priority thread is in middle of transaction. If we return here then successive poll outs from higher priority thread would be dropped, if we continue then we can lose first byte from higher priority thread. This point actually should never be reached as k_sleep will allow lower priority threads to execute and finish their operations. Safety counter was added just in case